fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rewrite/scalar/ReduceCastRule.java

```diff
@@ -164,32 +170,34 @@ public ScalarOperator optimizeDateTimeToDateCast(BinaryPredicateOperator operato
                 // Eg:cast dateTime(2021-12-28 00:00:00.0) to date(2021-12-28)
                 // Eg:cast dateTime(2021-12-28 00:00:00.1) to date(2021-12-27)
                 offset = 0;
-                targetLocalDateTime = bottomLocalDateTime.plusDays(offset);
-                newDate = ConstantOperator.createDate(targetLocalDateTime.truncatedTo(ChronoUnit.DAYS));
+                targetDateTime = bottomDateTime.plusDays(offset);
+                newDate = ConstantOperator.createDate(targetDateTime.truncatedTo(ChronoUnit.DAYS));
                 resultBinaryPredicateOperator =
                         new BinaryPredicateOperator(BinaryPredicateOperator.BinaryType.LE, castChild, newDate);
                 break;
             case LT:
                 // when the BinaryType is < ,cast dateTime to maximum date type；
                 // Eg:cast dateTime(2021-12-28 00:00:00.0) to date(2021-12-27)
                 // Eg:cast dateTime(2021-12-28 00:00:00.1) to date(2021-12-28)
-                if (originalDateTime.isEqual(bottomLocalDateTime)) {
+                if (originalDateTime.isEqual(bottomDateTime)) {
                     offset = -1;
                 } else {
                     offset = 0;
                 }
-                targetLocalDateTime = bottomLocalDateTime.plusDays(offset);
-                newDate = ConstantOperator.createDate(targetLocalDateTime.truncatedTo(ChronoUnit.DAYS));
+                targetDateTime = bottomDateTime.plusDays(offset);
+                newDate = ConstantOperator.createDate(targetDateTime.truncatedTo(ChronoUnit.DAYS));
                 resultBinaryPredicateOperator =
                         new BinaryPredicateOperator(BinaryPredicateOperator.BinaryType.LE, castChild, newDate);
                 break;
             case EQ:
                 // when the BinaryType is = ,cast dateTime to equivalent date type；
                 // Eg:cast dateTime(2021-12-28 00:00:00.0) to date(2021-12-28)
-                if (!originalDateTime.isEqual(bottomLocalDateTime)) {
-                    resultBinaryPredicateOperator = operator;
+                if (!originalDateTime.isEqual(bottomDateTime)) {
+                    // Since the format of cast(id_date as datetime) must be yyyy-MM-dd 00:00:00
+                    // yyyy-MM-dd 00:00:00 can never equal to yyyy-MM-dd ab:cd:ef (at least one of a/b/c/d/e/f is not zero)
+                    resultBinaryPredicateOperator = ConstantOperator.createBoolean(false);
```

Comment:
Can't return false direct, consider the id_date is NULL

Reply:
Ok, i got it